### PR TITLE
subsys: samples: doc: arch: Make perf subsystem

### PR DESCRIPTION
**Pull request by KushnerovMikhail**
Add profiling util based on periodic stack unwinding. Perf from Linux was taken as a reference.

Code is based on Sampling profiler (#29304) by @Jongy.

The operation of module is based on frame pointer usage and saving registers during interruption handling by zephyr.

General work description:
The unwinding function stay in timer as expiry function so is called during interruption handling. Thus the function have access to saved registers (program counter and frame pointer in particular) of the current thread and use it to unwind the thread stack.
Timer starting and results printing function are made as shell commands for conveniency.
Obtained samples can be translated into flamegraph using stackcollapse.py script.

Flame graph example, generated from `echo_server` sample:

graph

Kconfig optioins:
- `CONFIG_PERF` enables perf subsystem in compilation.
- `CONFIG_PERF_BUFFER_SIZE` specifies size of buffer, where samples are stored.

Kconfig requirements:
- `CONFIG_THREAD_STACK_INFO=y` provides `start` and `size` thread stack values.
- `CONFIG_SMP=n` SMP support is not implemented yet.
- `CONFIG_FRAME_POINTER=y` provide frame pointer.
- `CONFIG_SHELL=y` needs because subsystem provides shell commands.
- `CONFIG_RISCV=y` || `CONFIG_X86=y` - subsystem implemented only for x86 and riscv yet.

### CODE REVIEW

File: doc/services/debugging/perf.rst [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+   .. code-block:: bash
+
+      python zephyr/scripts/perf/stackcollapse.py perf_buf <build_dir>/zephyr/zephyr.elf > perf_buf.foolded
+
+   And turn into .svg FlameGraph (for example *graph.svg*)
+
+   .. code-block:: bash
+
+      ./FlameGraph/flamegraph.pl perf_buf.foolded > graph.svg
```
Comment (cfriedt) [inline doc/services/debugging/perf.rst:85]:
```suggestion
      ./FlameGraph/flamegraph.pl perf_buf.folded > graph.svg
```

File: doc/services/debugging/perf.rst [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      ....
+      00000000800163e0
+
+   Copy gotten output in some file, for example *perf_buf*.
+
+#. Translate perf samples by stackcollapse script to format, required by FlameGraph
+
+   .. code-block:: bash
+
+      python zephyr/scripts/perf/stackcollapse.py perf_buf <build_dir>/zephyr/zephyr.elf > perf_buf.foolded
```
Comment (cfriedt) [inline doc/services/debugging/perf.rst:79]:
```suggestion
      python zephyr/scripts/perf/stackcollapse.py perf_buf <build_dir>/zephyr/zephyr.elf > perf_buf.folded
```

**Review by cfriedt — DISMISSED** [review]
✨.Beautiful ✨.

**Review by nashif — CHANGES_REQUESTED** [review]
Please split the commit into multiple commits, i.e. 1 commit per arch, docs, adding the subsystem code and finally the sample.
Also, perf itself does not qualify as standalone subsystem. I suggest putting this under subsys/profiling/

File: doc/services/debugging/perf.rst (hunk 1 above)
Comment (KushnerovMikhail) [reply doc/services/debugging/perf.rst]: Thanks for the comment. Corrected it.

File: subsys/profiling/Kconfig [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# cOPYRIGHT (C) 2023 kns gROUP llc (yadro)
```
Comment (ycsin) [inline subsys/profiling/Kconfig:1]: This and other similar ones

```suggestion
# Copyright (c) 2023 KNS Group LLC (YADRO)
```

File: subsys/profiling/Kconfig [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig PROFILING
+	bool "Profiling tools"
+	help
+		Enable profiling tools, such as perf
```
Comment (ycsin) [inline subsys/profiling/Kconfig:8]: This and other Kconfig files, fix indent
```suggestion
	  Enable profiling tools, such as perf
```

File: subsys/profiling/perf/Kconfig [hunk 5]
```diff
@@ -0,0 +1,25 @@
+# cOPYRIGHT (C) 2023 kns gROUP llc (yadro)
+#
+# SPDX-License-Identifier: Apache-2.0
+
+config PROFILING_PERF
+	bool "Perf support"
+	default n
```
Comment (ycsin) [inline subsys/profiling/perf/Kconfig:7]: n is already default
```suggestion
```

File: subsys/profiling/perf/perf.c [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+					perf_data.buf + perf_data.idx,
+					CONFIG_PROFILING_PERF_BUFFER_SIZE - perf_data.idx);
+	}
+
+	if (trace_length != 0) {
+		perf_data.buf[perf_data.idx - 1] = trace_length;
+		perf_data.idx += trace_length;
+	} else {
+		--perf_data.idx;
+		printf("perf buf override!\n");
```
Comment (ycsin) [inline subsys/profiling/perf/perf.c:45]: should print to the shell, have a look at [k_timer_user_data_set](https://docs.zephyrproject.org/apidoc/latest/group__timer__apis.html#gadba1884961e790dd9c5d567de91cc7e2)

File: arch/riscv/core/perf.c [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+#include <zephyr/kernel.h>
+
+static bool valid_stack(uintptr_t addr, k_tid_t current)
+{
+	return current->stack_info.start <= addr &&
+		addr < current->stack_info.start + current->stack_info.size;
+}
+
+/*
+ * This function use frame poiters to unwind stack and get trace of return addresses.
```
Comment (ycsin) [inline arch/riscv/core/perf.c:16]:
```suggestion
 * This function use frame pointers to unwind stack and get trace of return addresses.
```

File: arch/riscv/core/perf.c [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+ * Return addresses are translated in corresponding function's names using .elf file.
+ * So we get function call trace
+ */
+size_t arch_perf_current_stack_trace(uintptr_t *buf, size_t size)
```
Comment (ycsin) [inline arch/riscv/core/perf.c:20]: Any chance that this and https://github.com/zephyrproject-rtos/zephyr/blob/main/arch/riscv/core/stacktrace.c can share some of the unwinding logics?

**Review by ycsin — COMMENTED** [review]
ARM64 supports stack unwinding as well. The docs can use some review from @kartben

File: arch/riscv/core/perf.c (hunk 8 above)
Comment (ycsin) [reply arch/riscv/core/perf.c:20]: See #73587

File: subsys/profiling/Kconfig (hunk 3 above)
Comment (KushnerovMikhail) [reply subsys/profiling/Kconfig]: I don't know how I let this happen. Thanks for the comment. Corrected it.

File: subsys/profiling/Kconfig (hunk 4 above)
Comment (KushnerovMikhail) [reply subsys/profiling/Kconfig]: Thanks for the comment. Corrected it.

File: subsys/profiling/perf/Kconfig (hunk 5 above)
Comment (KushnerovMikhail) [reply subsys/profiling/perf/Kconfig]: Corrected it.

File: subsys/profiling/perf/Kconfig [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+	depends on THREAD_STACK_INFO
+	depends on !SMP
+	depends on !OMIT_FRAME_POINTER
+	depends on OVERRIDE_FRAME_POINTER_DEFAULT
```
[the remark this answers is not in the document]
Comment (KushnerovMikhail) [reply subsys/profiling/perf/Kconfig]: Сhanged code to use a `CONFIG_FRAME_POINTER`.

File: subsys/profiling/perf/perf.c (hunk 6 above)
Comment (KushnerovMikhail) [reply subsys/profiling/perf/perf.c]: I had to create an additional handler thread because the shell_print does not work in interrupts.
I hope I understood correctly how you wanted me to use `k_timer_user_data_set`.

File: subsys/profiling/perf/perf.c [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+		k_timer_stop(&perf_data.timer);
+	}
+
+	if (perf_data.end_time <= k_uptime_get()) {
+		printf("perf done!\n");
+		k_timer_stop(&perf_data.timer);
+	}
+}
+
+struct shell;
```
[the remark this answers is not in the document]
Comment (KushnerovMikhail) [reply subsys/profiling/perf/perf.c]: Really not necessary. Thanks for the comment.

File: subsys/profiling/perf/perf.c [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+	if (perf_data.end_time == 0) {
+		k_timer_init(&perf_data.timer, perf_tracer, NULL);
+	}
+
+	perf_data.end_time = k_uptime_get() + strtoll(argv[1], NULL, 10);
+	int64_t period = 1000000000 / strtoll(argv[2], NULL, 10);
+
+	k_timer_start(&perf_data.timer, K_NO_WAIT, K_NSEC(period));
+
+	printf("Enabled perf\n");
```
[the remark this answers is not in the document]
Comment (KushnerovMikhail) [reply subsys/profiling/perf/perf.c]: Corrected it

File: subsys/profiling/perf/perf.c [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+	return 0;
+}
+
+int cmd_perf_print(const struct shell *sh, size_t argc, char **argv)
+{
+	printf("Perf buf length %zu\n", perf_data.idx);
```
[the remark this answers is not in the document]
Comment (KushnerovMikhail) [reply subsys/profiling/perf/perf.c]: Corrected it

File: subsys/profiling/perf/perf.c [hunk 13]
```diff
@@ -0,0 +1,98 @@
...
+{
+	ARG_UNUSED(argc);
+	ARG_UNUSED(argv);
+
+	shell_print(sh, "perfy");
+	return 0;
+}
+
+SHELL_STATIC_SUBCMD_SET_CREATE(m_sub_perf,
+	SHELL_CMD_ARG(record, NULL, "Start recording for X ms on Y Hz", cmd_perf_record, 3, 0),
```
[the remark this answers is not in the document]
Comment (KushnerovMikhail) [reply subsys/profiling/perf/perf.c]: Your help string is really better then mine. Thanks for help

File: arch/riscv/core/perf.c (hunk 7 above)
Comment (KushnerovMikhail) [reply arch/riscv/core/perf.c]: Corrected it.

File: subsys/profiling/perf/perf.c (hunk 6 above)
Comment (ycsin) [reply subsys/profiling/perf/perf.c]: Yeah I didn't know that the timer handler is called directly from an interrupt context, I managed to use a workqueue instead of a dedicated thread in https://github.com/zephyrproject-rtos/zephyr/commit/501c9093245fbc70df69ab3a07faa0055455e2dc (there are some other changes, as I was testing #73587)

File: arch/riscv/core/perf.c [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+	/*
+	 * In riscv (arch/riscv/core/isr.S) ra, ip($mepc) and fp($s0) are saved
+	 * at the beginning of _isr_wrapper in order, specified by z_arch_esf_t.
+	 * Then, before calling interruption handler, core switch $sp to
+	 * _current_cpu->irq_stack and save $sp with offset -16 on irq stack
+	 *
+	 * The following lines lines do the reverse things to get ra, ip anf fp
+	 * from thread stack
+	 */
+	const z_arch_esf_t * const esf =
```
Comment (nashif) [inline arch/riscv/core/perf.c:36]: this was recently deprecated, can you please update? @ycsin fyi

Comment (ycsin) [reply arch/riscv/core/perf.c:36]: > this was recently deprecated, can you please update? @ycsin fyi

I think it makes sense for this subsys to use a generic arch level API provided in #73587

**Comment by KushnerovMikhail** [comment]
Added implementation for x86_32. For it I had to replace saving %edi with %ebp in interruption context saving code. Seems like it is not ruin anything.

Also I changed a little riscv stack unwinding code to avoid bug, when interruption occurs in function prologue or epilogue. The bug was that in epilogue/prologue fp is equal to pointer of previous function frame, and it looks like second from top function is missed.

graph
\\/ 
graph
You can see that on the first graph `sys_clock_cycle_get_32` is called not only from `z_impl_k_busy_wait`, that is wrong.

Rewrite subsystem with workqueue thread instead of events

File: arch/riscv/core/perf.c (hunk 14 above)
Comment (KushnerovMikhail) [reply arch/riscv/core/perf.c]: Fixed deprecated code

Comment (ycsin) [reply arch/riscv/core/perf.c]: I think you are right, similar handling exists in Linux's implementation

**Comment by ycsin** [comment]
> I am thinking about using API provided in #73587, but there are two issues, I have with it:
> 
> * Riscv implementation do not check case when interruption occurred in function where $ra is not saved in stack frame (it can be if function does not call any other function). Seems like it leads to undefined behaviour (in this case `frame->ra=$fp`, `frame->fp=<undefined>`).
> * In riscv context saving code is same for exceptions and interruptions so during interruption registers are saved according to `struct arch_esf`. But for x86 there is separation between context saving during interruption and exception, and registers saved during interruption does not matches with `struct arch_esf`. And perf unwind stack during interruption.

I think from the project perspective it makes sense for applications/kernel to use the same stack-unwinding function so that there's less code duplication and easier to maintain in the long run, which is what the `arch_stack_walk()` in #73587 is trying to achieve.

I can probably fix the issue that you mentioned in RISC-V. However, I do not know how stack unwinding works in x86/arm64, but I can massage their existing implementations to use the `arch_stack_walk()`, hopefully that makes it a little easier for you to update this PR to use `arch_stack_walk()`, does this work for you?

**Review by ycsin — COMMENTED** [review]
Should add a new entry to https://github.com/zephyrproject-rtos/zephyr/blob/main/MAINTAINERS.yml

File: arch/x86/core/ia32/intstub.S [hunk 15]
```diff
@@ -112,11 +112,12 @@ SECTION_FUNC(PINNED_TEXT, _interrupt_enter)
 	 * EAX = isr_param, EDX = isr
 	 */
 
-	/* Push EDI as we will use it for scratch space.
+	/* Push EBP as we will use it for scratch space.
```
Comment (ycsin) [inline arch/x86/core/ia32/intstub.S:115]: The changes to this file could probably be in another commit / PR to document this change better

**Comment by teburd** [comment]
Now we're cooking

**Comment by nashif** [comment]
looks good, will take another look so we can get it ready when main branch opens

**Review by nashif — APPROVED** [review]

**Review by teburd — APPROVED** [review]
